### DISCUSSION TITLE
`parallelAlgorithm` behavior documentation inaccuracy?

### DISCUSSION THREAD

**Discussion by siewers** [xunit/xunit#2958] [Question] [answered]
I've just updated to xUnit 2.8.
In the documentation concerning [algorithms](https://xunit.net/docs/running-tests-in-parallel#algorithms), when running tests in parallel, it's stated that the `parallelAlgorithm` has the following behavior:
>`conservative`
>This will only start as many tests as _your max parallel threads setting_.

>`aggressive`
>This is the original parallelism algorithm, which starts as many tests as possible, _regardless of your max parallel threads setting_, and uses a [`SynchronizationContext`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.synchronizationcontext) to limit the number of things that are running at any given time.

However, I've been doing some testing, and it _seems_ the behavior is in reverse.

The test:
```cs
using Xunit;

namespace ParallelAlgorithmTests;

public class UnitTest1
{
    public static IEnumerable<object[]> TestCases()
    {
        return Enumerable.Range(0, 1000).Select(i => new object[] { i });
    }

    [Theory]
    [MemberData(nameof(TestCases))]
    public async Task Test(int index)
    {
        await ContextSwitchHelper.Switch(10000);
    }
}

public static class ContextSwitchHelper
{
    public static async Task Switch(int times)
    {
        for (var i = 0; i < times; i++)
        {
            await Task.Yield();
        }
    }
}
```

**`aggressive`**
```bash
#Using 1 thread
dotnet test --no-build -- xUnit.ParallelAlgorithm=aggressive xUnit.MaxParallelThreads=1
#Passed!  - Failed:     0, Passed:  1000, Skipped:     0, Total:  1000, Duration: 9 s

#Using 12 threads
dotnet test --no-build -- xUnit.ParallelAlgorithm=aggressive xUnit.MaxParallelThreads=12
#Passed!  - Failed:     0, Passed:  1000, Skipped:     0, Total:  1000, Duration: 50 s
```

**`conservative`**
```bash
#Using 1 thread
dotnet test --no-build -- xUnit.ParallelAlgorithm=conservative xUnit.MaxParallelThreads=1
#Passed!  - Failed:     0, Passed:  1000, Skipped:     0, Total:  1000, Duration: 4 s

#Using 12 threads
dotnet test --no-build -- xUnit.ParallelAlgorithm=conservative xUnit.MaxParallelThreads=12
#Passed!  - Failed:     0, Passed:  1000, Skipped:     0, Total:  1000, Duration: 4 s
```

The `MaxParallelThreads` parameter affects the behavior using the `aggressive` algorithm, while the `conservative` algorithm seems unaffected.

I didn't want to create this as a bug yet since there may be obvious reasons for this behavior that I do not understand, but according to my tests, the documentation is inaccurate.

It may also be that `conservative` _does_ take the `MaxParallelThreads` into account, just doing a much better job, so the number doesn't affect the run at all (that seems plausible given it's faster than the "fastest" run using `aggressive`).

However, the claim in the documentation that `MaxParallelThreads` is ignored with `aggressive` seems incorrect.

So, is the documentation wrong, or am I? :)

**Comment by bradwilson** [comment, marked as answer]
Intra-assembly parallelism in xUnit.net is based on around test collections. Tests in the same collection don't run in parallel against each other; tests in different collections do.

By default, each test class is a test collection (and with the default tools, that's the smallest boundary of tests in the same collection), so tests in the same class will never run in parallel against each other. In order to test this appropriately, what you need is 10,000 test classes (each with one test).
